contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,995 @@
+     *             only step 1 is skipped. (It is important to note that liquidity needs to exist in the configured
+     *             strategy tick ranges in order for the swap to be possible) Step 3 mints new liquidity position
+     *             instead of adding to an existing one.
+     * Scenario 3: When there is strategy's liquidity in the pool all 3 steps are taken
+     *
+     *
+     * Exact _amountToSwap, _minTokenReceived & _swapWeth parameters shall be determined by simulating the
+     * transaction off-chain. The strategy checks that after the swap the share of the tokens is in the
+     * expected ranges.
+     *
+     * @param _amountToSwap The amount of the token to swap
+     * @param _swapWeth Swap using WETH when true, use OETHb when false
+     * @param _minTokenReceived Slippage check -> minimum amount of token expected in return
+     */
+    function rebalance(
+        uint256 _amountToSwap,
+        bool _swapWeth,
+        uint256 _minTokenReceived
+    ) external nonReentrant onlyGovernorOrStrategist {
+        /**
+         * Would be nice to check if there is any total liquidity in the pool before performing this swap
+         * but there is no easy way to do that in UniswapV3:
+         * - clPool.liquidity() -> only liquidity in the active tick
+         * - asset[1&2].balanceOf(address(clPool)) -> will include uncollected tokens of LP providers
+         *   after their liquidity position has been decreased
+         */
+
+        /**
+         * When rebalance is called for the first time there is no strategy's
+         * liquidity in the pool yet. The full liquidity removal is thus skipped.
+         */
+        if (tokenId != 0) {
+            _removeLiquidity(1e18);
+        }
+        // in some cases we will just want to add liquidity and not issue a swap to move the
+        // active trading position within the pool
+        if (_amountToSwap > 0) {
+            _swapToDesiredPosition(_amountToSwap, _swapWeth, _minTokenReceived);
+        }
+        // calling check liquidity early so we don't get unexpected errors when adding liquidity
+        // in the later stages of this function
+        _checkForExpectedPoolPrice();
+
+        _addLiquidity();
+        _checkForExpectedPoolPrice();
+
+        // revert if protocol insolvent
+        _solvencyAssert();
+    }
+
+    /**
+     * Checks that the protocol is solvent, protecting from a rogue Strategist / Guardian that can
+     * keep rebalancing the pool in both directions making the protocol lose a tiny amount of
+     * funds each time.
+     *
+     * Protocol must be at least SOLVENCY_THRESHOLD (99,8 %) backed in order for the rebalances to
+     * function.
+     */
+    function _solvencyAssert() internal view {
+        uint256 _totalVaultValue = IVault(vaultAddress).totalValue();
+        uint256 _totalOethbSupply = IERC20(OETHb).totalSupply();
+
+        if (
+            _totalVaultValue.divPrecisely(_totalOethbSupply) <
```

Comment:
This code here is fine for this purpose - no need to change it.

But it does do rounding, and it does depend on one of the values being greater than zero. You could instead write something like:

`a * 0.998 ethers < b * 1 ether`

Again, no need to change, just might come in handy some day.

Reply:
Hmm that is a great point thanks. I haven't though about it that way yet. In this case that would be
```
totalOethbSupply * SOLVENCY_THRESHOLD < totalVaultValue * 1 ether
```

For me though it is very unnatural way of expressing code. So a comment explaining what expression it is replacing would probably be necessary

Reply:
Yeah, I think this style is a solution to a problem that this particular code doesn't have.